idle_sweep_test.go

```diff
@@ -0,0 +1,239 @@
+
+	"github.com/uber/tchannel-go/raw"
+	"github.com/uber/tchannel-go/testutils"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/net/context"
+)
+
+func numConnections(ch *Channel) int {
+	rootPeers := ch.RootPeers().Copy()
+	count := 0
+
+	for _, peer := range rootPeers {
+		in, out := peer.NumConnections()
+
+		ch.Logger().Infof("numConnections IN: %d  OUT: %d", in, out)
+		count += in + out
+	}
+
+	ch.Logger().Infof("numConnections TOTAL: %d", count)
+	return count
+}
+
+// Validates that inbound idle connections are dropped.
+func TestServerBasedSweep(t *testing.T) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+
+	serverTicker := testutils.NewFakeTicker()
+	clock := testutils.NewStubClock(time.Now())
+
+	serverOpts := testutils.NewOpts().
+		SetTimeTicker(serverTicker.New).
+		SetIdleCheckInterval(30 * time.Second).
+		SetMaxIdleTime(3 * time.Minute).
+		SetTimeNow(clock.Now).
+		NoRelay()
+
+	clientOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now)
+
+	testutils.WithTestServer(t, serverOpts, func(ts *testutils.TestServer) {
+		ts.RegisterFunc("echo", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+			return &raw.Res{}, nil
+		})
+
+		client := ts.NewClient(clientOpts)
+		raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+
+		// Both server and client now have an active connection. After 3 minutes they
+		// should be cleared out by the idle sweep.
+		for i := 0; i < 2; i++ {
+			clock.Elapse(1 * time.Minute)
+
+			serverTicker.Tick()
+			time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+			assert.Equal(t, 1, numConnections(ts.Server()))
+			assert.Equal(t, 1, numConnections(client))
+		}
+
+		clock.Elapse(90 * time.Second)
+		serverTicker.Tick()
+
+		// Allow some time for the disconnect to propagate
+		time.Sleep(testutils.Timeout(10 * time.Millisecond))
+
+		assert.Equal(t, 0, numConnections(ts.Server()))
+		assert.Equal(t, 0, numConnections(client))
+	})
+}
+
+// Validates that outbound idle connections are dropped.
+func TestClientBasedSweep(t *testing.T) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+
+	clientTicker := testutils.NewFakeTicker()
+	clock := testutils.NewStubClock(time.Now())
+
+	serverOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now).
+		NoRelay()
+
+	clientOpts := testutils.NewOpts().
+		SetTimeNow(clock.Now).
+		SetTimeTicker(clientTicker.New).
+		SetMaxIdleTime(3 * time.Minute).
+		SetIdleCheckInterval(30 * time.Second)
+
+	testutils.WithTestServer(t, serverOpts, func(ts *testutils.TestServer) {
+		ts.RegisterFunc("echo", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+			return &raw.Res{}, nil
+		})
+
+		client := ts.NewClient(clientOpts)
+		raw.Call(ctx, client, ts.HostPort(), ts.ServiceName(), "echo", nil, nil)
+
+		// Both server and client now have an active connection. After 3 minutes they
+		// should be cleared out by the idle sweep.
+		clientTicker.Tick()
+		time.Sleep(testutils.Timeout(10 * time.Millisecond))
```

Comment:
why do we need sleeps here? if anything we should have a `testutils.WaitFor`, but I'm strongly against sleeps in tests, they end up being flaky or cause the tests to slow down

Reply:
I agree, this was a tough one. The call to `checkIdleConnections` happens in a goroutine and we have no synchronization channel to know when it completed. There is additionally some small delay between disconnecting the client-side until the server-side receives and processes the disconnect, even though they are both happening in localhost.

I changed to use `testutils.WaitFor` but I'm open to other suggestions.